File: Examples/ObjCSampleApp/ObjCSampleApp/MainViewController.m
#import "MainViewController.h"

#include <stdio.h>
#include <string.h>
#include <unistd.h>

@import ConsoleDock;
@import ConsoleDockCore;

static void SampleAppLogInfo(NSString *message)
{
    NSString *formattedMessage = [NSString stringWithFormat:@"[sample app logger] %@", message];
    NSLog(@"%@", formattedMessage);
    [CDKConsoleDock info:formattedMessage];
}

static NSString *const SampleAccessibilityStatusIdentifier = @"objc-sample.status";

static NSString *SampleAccessibilityButtonIdentifier(NSString *title)
{
    NSCharacterSet *separatorSet = [[NSCharacterSet alphanumericCharacterSet] invertedSet];
    NSArray<NSString *> *rawComponents = [title.lowercaseString componentsSeparatedByCharactersInSet:separatorSet];
    NSMutableArray<NSString *> *components = [NSMutableArray array];
    for (NSString *component in rawComponents) {
        if (component.length > 0) {
            [components addObject:component];
        }
    }
    return [@"objc-sample." stringByAppendingString:[components componentsJoinedByString:@"-"]];
}

@interface MainViewController ()

@property (nonatomic, strong) UILabel *statusLabel;
@property (nonatomic) NSInteger counter;

@end

@implementation MainViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.title = @"ConsoleDock ObjC";
    self.view.backgroundColor = [UIColor colorWithWhite:0.96 alpha:1.0];
    [self configureView];
    [self updateStatus:@"ConsoleDock started. Tap CD or Show Console."];
}

- (void)configureView
{
    UILabel *headingLabel = [[UILabel alloc] init];
    headingLabel.text = @"ConsoleDock Objective-C Sample";
    headingLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleTitle2];
    headingLabel.textColor = UIColor.blackColor;
    headingLabel.numberOfLines = 0;

    UILabel *bodyLabel = [[UILabel alloc] init];
    bodyLabel.text = @"Generate Objective-C, C stdio, direct descriptor writes, and NSLog output, then open the in-app console.";
    bodyLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
    bodyLabel.textColor = [UIColor colorWithWhite:0.25 alpha:1.0];
    bodyLabel.numberOfLines = 0;

    self.statusLabel = [[UILabel alloc] init];
    self.statusLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleFootnote];
    self.statusLabel.textColor = [UIColor colorWithWhite:0.28 alpha:1.0];
    self.statusLabel.numberOfLines = 0;
    self.statusLabel.accessibilityIdentifier = SampleAccessibilityStatusIdentifier;

    UIStackView *stackView = [[UIStackView alloc] initWithArrangedSubviews:@[
        headingLabel,
        bodyLabel,
        [self makeButtonWithTitle:@"Show Console" action:@selector(showConsole)],
        [self makeButtonWithTitle:@"Log diagnostics" action:@selector(logDiagnostics)],
        [self makeButtonWithTitle:@"App logger sink" action:@selector(logAppLoggerSink)],
        [self makeButtonWithTitle:@"CDKConsoleDock info" action:@selector(logNativeInfo)],
        [self makeButtonWithTitle:@"CDKConsoleDock error" action:@selector(logNativeError)],
        [self makeButtonWithTitle:@"CDKConsoleDock fault" action:@selector(logNativeFault)],
        [self makeButtonWithTitle:@"printf stdout" action:@selector(logPrintf)],
        [self makeButtonWithTitle:@"fprintf stderr" action:@selector(logStderr)],
        [self makeButtonWithTitle:@"write stdout" action:@selector(logWriteStdout)],
        [self makeButtonWithTitle:@"write stderr" action:@selector(logWriteStderr)],
        [self makeButtonWithTitle:@"NSLog" action:@selector(logNSLog)],
        [self makeButtonWithTitle:@"Clear ConsoleDock Entries" action:@selector(clearEntries)],
        [self makeButtonWithTitle:@"Stop ConsoleDock" action:@selector(stopConsoleDock)],
        [self makeButtonWithTitle:@"Start ConsoleDock" action:@selector(startConsoleDock)],
        self.statusLabel
    ]];
    stackView.translatesAutoresizingMaskIntoConstraints = NO;
    stackView.axis = UILayoutConstraintAxisVertical;
    stackView.alignment = UIStackViewAlignmentFill;
    stackView.spacing = 12;

    UIScrollView *scrollView = [[UIScrollView alloc] init];
    scrollView.translatesAutoresizingMaskIntoConstraints = NO;
    scrollView.alwaysBounceVertical = YES;
    [scrollView addSubview:stackView];
    [self.view addSubview:scrollView];

    UILayoutGuide *safeArea = self.view.safeAreaLayoutGuide;
    [NSLayoutConstraint activateConstraints:@[
        [scrollView.leadingAnchor constraintEqualToAnchor:safeArea.leadingAnchor],
        [scrollView.trailingAnchor constraintEqualToAnchor:safeArea.trailingAnchor],
        [scrollView.topAnchor constraintEqualToAnchor:safeArea.topAnchor],
        [scrollView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor],

        [stackView.leadingAnchor constraintEqualToAnchor:scrollView.contentLayoutGuide.leadingAnchor constant:20],
        [stackView.trailingAnchor constraintEqualToAnchor:scrollView.contentLayoutGuide.trailingAnchor constant:-20],
        [stackView.topAnchor constraintEqualToAnchor:scrollView.contentLayoutGuide.topAnchor constant:24],
        [stackView.bottomAnchor constraintEqualToAnchor:scrollView.contentLayoutGuide.bottomAnchor constant:-24],
        [stackView.widthAnchor constraintEqualToAnchor:scrollView.frameLayoutGuide.widthAnchor constant:-40]
    ]];
}

- (UIButton *)makeButtonWithTitle:(NSString *)title action:(SEL)action
{
    UIButton *button = [UIButton buttonWithType:UIButtonTypeSystem];
    [button setTitle:title forState:UIControlStateNormal];
    button.accessibilityIdentifier = SampleAccessibilityButtonIdentifier(title);
    button.titleLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
    button.titleLabel.adjustsFontForContentSizeCategory = YES;
    button.backgroundColor = UIColor.whiteColor;
    button.layer.borderColor = [UIColor colorWithWhite:0.78 alpha:1.0].CGColor;
    button.layer.borderWidth = 1;
    button.layer.cornerRadius = 8;
    button.contentEdgeInsets = UIEdgeInsetsMake(12, 14, 12, 14);
    [button addTarget:self action:action forControlEvents:UIControlEventTouchUpInside];
    return button;
}

- (NSString *)nextMessageWithPrefix:(NSString *)prefix
{
    self.counter += 1;
    return [NSString stringWithFormat:@"%@ #%ld token=objc-secret-%ld",
                                      prefix,
                                      (long)self.counter,
                                      (long)self.counter];
}

- (void)updateStatus:(NSString *)message
{
    CDKDiagnostics *diagnostics = [CDKConsoleDock diagnostics];
    self.statusLabel.text = [NSString stringWithFormat:@"%@\nRunning: %@  Stored entries: %lu  "
                                                       @"stdout: %@  stderr: %@",
                                                       message,
                                                       diagnostics.isRunning ? @"YES" : @"NO",
                                                       (unsigned long)diagnostics.entryCount,
                                                       diagnostics.captureStandardOutput ? @"YES" : @"NO",
                                                       diagnostics.captureStandardError ? @"YES" : @"NO"];
}

- (void)updateStatusAfterCapture:(NSString *)message
{
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.15 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self updateStatus:message];
    });
}

- (void)showConsole
{
    [CDKConsoleDockUIKit showConsole];
    [self updateStatus:@"Requested ConsoleDock panel."];
}

- (void)logDiagnostics
{
    CDKDiagnostics *diagnostics = [CDKConsoleDock diagnostics];
    NSString *message = [NSString stringWithFormat:@"objc diagnostics running=%@ entries=%lu "
                                                   @"stdout=%@ stderr=%@ limits=%lu/%lu "
                                                   @"redacted=%lu truncated=%lu partial=%lu",
                                                   diagnostics.isRunning ? @"YES" : @"NO",
                                                   (unsigned long)diagnostics.entryCount,
                                                   diagnostics.captureStandardOutput ? @"YES" : @"NO",
                                                   diagnostics.captureStandardError ? @"YES" : @"NO",
                                                   (unsigned long)diagnostics.maximumEntries,
                                                   (unsigned long)diagnostics.maximumMessageLength,
                                                   (unsigned long)diagnostics.redactedEntryCount,
                                                   (unsigned long)diagnostics.truncatedEntryCount,
                                                   (unsigned long)diagnostics.partialEntryCount];
    [CDKConsoleDock info:message];
    [self updateStatus:@"Wrote ConsoleDock diagnostics."];
}

- (void)logAppLoggerSink
{
    SampleAppLogInfo([self nextMessageWithPrefix:@"objc app logger sink"]);
    [self updateStatusAfterCapture:@"Wrote app logger sink."];
}

- (void)logNativeInfo
{
    NSString *message = [self nextMessageWithPrefix:@"objc native info"];
    [CDKConsoleDock info:message];
    [self updateStatus:@"Wrote CDKConsoleDock info."];
}

- (void)logNativeError
{
    NSString *message = [self nextMessageWithPrefix:@"objc native error"];
    [CDKConsoleDock error:message];
    [self updateStatus:@"Wrote CDKConsoleDock error."];
}

- (void)logNativeFault
{
    NSString *message = [self nextMessageWithPrefix:@"objc native fault"];
    [CDKConsoleDock fault:message];
    [self updateStatus:@"Wrote CDKConsoleDock fault."];
}

- (void)logPrintf
{
    NSString *message = [self nextMessageWithPrefix:@"objc printf stdout"];
    printf("%s\n", message.UTF8String);
    fflush(stdout);
    [self updateStatusAfterCapture:@"Wrote printf stdout."];
}

- (void)logStderr
{
    NSString *message = [self nextMessageWithPrefix:@"objc fprintf stderr"];
    fprintf(stderr, "%s\n", message.UTF8String);
    fflush(stderr);
    [self updateStatusAfterCapture:@"Wrote fprintf stderr."];
}

- (void)logWriteStdout
{
    NSString *message = [[self nextMessageWithPrefix:@"objc write stdout"] stringByAppendingString:@"\n"];
    const char *bytes = message.UTF8String;
    write(STDOUT_FILENO, bytes, strlen(bytes));
    [self updateStatusAfterCapture:@"Wrote direct stdout."];
}

- (void)logWriteStderr
{
    NSString *message = [[self nextMessageWithPrefix:@"objc write stderr"] stringByAppendingString:@"\n"];
    const char *bytes = message.UTF8String;
    write(STDERR_FILENO, bytes, strlen(bytes));
    [self updateStatusAfterCapture:@"Wrote direct stderr."];
}

- (void)logNSLog
{
    NSLog(@"%@", [self nextMessageWithPrefix:@"objc NSLog output"]);
    [self updateStatusAfterCapture:@"Wrote NSLog."];
}

- (void)clearEntries
{
    [CDKConsoleDock clearEntries];
    [self updateStatus:@"Cleared ConsoleDock entries."];
}

- (void)stopConsoleDock
{
    [CDKConsoleDockUIKit stop];
    [self updateStatus:@"Stopped ConsoleDock."];
}

- (void)startConsoleDock
{
    CDKConfiguration *configuration = [CDKConfiguration defaultConfiguration];
    configuration.maximumEntries = 500;
    configuration.maximumMessageLength = 4096;
    configuration.captureStandardOutput = YES;
    configuration.captureStandardError = YES;
    configuration.showsFloatingButton = YES;
    configuration.allowsReleaseBuilds = NO;

    NSError *error = nil;
    CDKStartResult result = [CDKConsoleDockUIKit startWithConfiguration:configuration error:&error];
    if (result == CDKStartResultFailed) {
        [self updateStatus:[NSString stringWithFormat:@"Failed to start ConsoleDock: %@",
                                                      error.localizedDescription ?: @"Unknown error"]];
        return;
    }

    [self updateStatus:@"Started ConsoleDock."];
}

@end

File: Examples/SwiftSampleApp/SwiftSampleApp/MainViewController.swift
import ConsoleDock
import UIKit

private enum SampleAppLog {
    private static let consoleDockForwarder = ConsoleDock.LogForwarder(category: "sample app logger")

    static func info(_ message: String) {
        let formattedMessage = "[sample app logger] \(message)"
        print(formattedMessage)
        consoleDockForwarder.info(message)
    }
}

private enum SampleAccessibilityIdentifiers {
    static let status = "swift-sample.status"

    static func button(_ title: String) -> String {
        let components =
            title
            .lowercased()
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .filter { !$0.isEmpty }
        return "swift-sample.\(components.joined(separator: "-"))"
    }
}

final class MainViewController: UIViewController {
    private let statusLabel = UILabel()
    private var counter = 0

    override func viewDidLoad() {
        super.viewDidLoad()
        title = "ConsoleDock Sample"
        view.backgroundColor = UIColor(white: 0.96, alpha: 1)
        configureView()
        updateStatus("ConsoleDock started. Tap CD or Show Console.")
    }

    private func configureView() {
        let headingLabel = UILabel()
        headingLabel.text = "ConsoleDock Swift Sample"
        headingLabel.font = .preferredFont(forTextStyle: .title2)
        headingLabel.textColor = .black
        headingLabel.numberOfLines = 0

        let bodyLabel = UILabel()
        bodyLabel.text = "Generate stdout, stderr, and native ConsoleDock logs, then open the in-app console."
        bodyLabel.font = .preferredFont(forTextStyle: .body)
        bodyLabel.textColor = UIColor(white: 0.25, alpha: 1)
        bodyLabel.numberOfLines = 0

        statusLabel.font = .preferredFont(forTextStyle: .footnote)
        statusLabel.textColor = UIColor(white: 0.28, alpha: 1)
        statusLabel.numberOfLines = 0
        statusLabel.accessibilityIdentifier = SampleAccessibilityIdentifiers.status

        let stackView = UIStackView(arrangedSubviews: [
            headingLabel,
            bodyLabel,
            makeButton(title: "Show Console", action: #selector(showConsole)),
            makeButton(title: "Hide Floating Button", action: #selector(hideFloatingButton)),
            makeButton(title: "Show Floating Button", action: #selector(showFloatingButton)),
            makeButton(title: "Log diagnostics", action: #selector(logDiagnostics)),
            makeButton(title: "App logger sink", action: #selector(logAppLoggerSink)),
            makeButton(title: "ConsoleDock.info", action: #selector(logNativeInfo)),
            makeButton(title: "ConsoleDock.error", action: #selector(logNativeError)),
            makeButton(title: "ConsoleDock.fault", action: #selector(logNativeFault)),
            makeButton(title: "print stdout", action: #selector(logPrint)),
            makeButton(title: "printf stdout", action: #selector(logPrintf)),
            makeButton(title: "fprintf stderr", action: #selector(logStderr)),
            makeButton(title: "NSLog", action: #selector(logNSLog)),
            makeButton(title: "Clear ConsoleDock Entries", action: #selector(clearEntries)),
            makeButton(title: "Stop ConsoleDock", action: #selector(stopConsoleDock)),
            makeButton(title: "Start ConsoleDock", action: #selector(startConsoleDock)),
            statusLabel
        ])
        stackView.translatesAutoresizingMaskIntoConstraints = false
        stackView.axis = .vertical
        stackView.alignment = .fill
        stackView.spacing = 12

        let scrollView = UIScrollView()
        scrollView.translatesAutoresizingMaskIntoConstraints = false
        scrollView.alwaysBounceVertical = true
        scrollView.addSubview(stackView)
        view.addSubview(scrollView)

        NSLayoutConstraint.activate([
            scrollView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
            scrollView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
            scrollView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            scrollView.bottomAnchor.constraint(equalTo: view.bottomAnchor),

            stackView.leadingAnchor.constraint(equalTo: scrollView.contentLayoutGuide.leadingAnchor, constant: 20),
            stackView.trailingAnchor.constraint(equalTo: scrollView.contentLayoutGuide.trailingAnchor, constant: -20),
            stackView.topAnchor.constraint(equalTo: scrollView.contentLayoutGuide.topAnchor, constant: 24),
            stackView.bottomAnchor.constraint(equalTo: scrollView.contentLayoutGuide.bottomAnchor, constant: -24),
            stackView.widthAnchor.constraint(equalTo: scrollView.frameLayoutGuide.widthAnchor, constant: -40)
        ])
    }

    private func makeButton(title: String, action: Selector) -> UIButton {
        let button = UIButton(type: .system)
        button.setTitle(title, for: .normal)
        button.accessibilityIdentifier = SampleAccessibilityIdentifiers.button(title)
        button.titleLabel?.font = .preferredFont(forTextStyle: .body)
        button.titleLabel?.adjustsFontForContentSizeCategory = true
        button.backgroundColor = UIColor(white: 1, alpha: 1)
        button.layer.borderColor = UIColor(white: 0.78, alpha: 1).cgColor
        button.layer.borderWidth = 1
        button.layer.cornerRadius = 8
        button.contentEdgeInsets = UIEdgeInsets(top: 12, left: 14, bottom: 12, right: 14)
        button.addTarget(self, action: action, for: .touchUpInside)
        return button
    }

    private func nextMessage(_ prefix: String) -> String {
        counter += 1
        return "\(prefix) #\(counter) token=sample-secret-\(counter)"
    }

    private func updateStatus(_ message: String) {
        let diagnostics = ConsoleDock.diagnostics
        let status = [
            "Running: \(diagnostics.isRunning)",
            "Stored entries: \(diagnostics.entryCount)",
            "stdout: \(diagnostics.capturesStandardOutput)",
            "stderr: \(diagnostics.capturesStandardError)"
        ].joined(separator: "  ")
        statusLabel.text = "\(message)\n\(status)"
    }

    private func updateStatusAfterCapture(_ message: String) {
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.15) { [weak self] in
            self?.updateStatus(message)
        }
    }

    @objc private func showConsole() {
        ConsoleDock.showConsole()
        updateStatus("Requested ConsoleDock panel.")
    }

    @objc private func hideFloatingButton() {
        ConsoleDock.hideFloatingButton()
        updateStatus("Hid ConsoleDock floating button.")
    }

    @objc private func showFloatingButton() {
        ConsoleDock.showFloatingButton()
        updateStatus("Showed ConsoleDock floating button.")
    }

    @objc private func logDiagnostics() {
        let diagnostics = ConsoleDock.diagnostics
        let message = [
            "diagnostics",
            "running=\(diagnostics.isRunning)",
            "entries=\(diagnostics.entryCount)",
            "stdout=\(diagnostics.capturesStandardOutput)",
            "stderr=\(diagnostics.capturesStandardError)",
            "limits=\(diagnostics.maximumEntries)/\(diagnostics.maximumMessageLength)",
            "redacted=\(diagnostics.redactedEntryCount)",
            "truncated=\(diagnostics.truncatedEntryCount)",
            "partial=\(diagnostics.partialEntryCount)"
        ].joined(separator: " ")
        ConsoleDock.info(message)
        updateStatus("Wrote ConsoleDock diagnostics.")
    }

    @objc private func logAppLoggerSink() {
        SampleAppLog.info(nextMessage("app logger sink"))
        fflush(stdout)
        updateStatusAfterCapture("Wrote app logger sink.")
    }

    @objc private func logNativeInfo() {
        let message = nextMessage("native info")
        ConsoleDock.info(message)
        updateStatus("Wrote ConsoleDock.info.")
    }

    @objc private func logNativeError() {
        let message = nextMessage("native error")
        ConsoleDock.error(message)
        updateStatus("Wrote ConsoleDock.error.")
    }

    @objc private func logNativeFault() {
        let message = nextMessage("native fault")
        ConsoleDock.fault(message)
        updateStatus("Wrote ConsoleDock.fault.")
    }

    @objc private func logPrint() {
        print(nextMessage("print stdout"))
        fflush(stdout)
        updateStatusAfterCapture("Wrote print stdout.")
    }

    @objc private func logPrintf() {
        nextMessage("printf stdout").withCString { message in
            SampleLogPrintf(message)
        }
        updateStatusAfterCapture("Wrote printf stdout.")
    }

    @objc private func logStderr() {
        nextMessage("fprintf stderr").withCString { message in
            SampleLogFprintfStderr(message)
        }
        updateStatusAfterCapture("Wrote fprintf stderr.")
    }

    @objc private func logNSLog() {
        NSLog("%@", nextMessage("NSLog output"))
        updateStatusAfterCapture("Wrote NSLog.")
    }

    @objc private func clearEntries() {
        ConsoleDock.clear()
        updateStatus("Cleared ConsoleDock entries.")
    }

    @objc private func stopConsoleDock() {
        ConsoleDock.stop()
        updateStatus("Stopped ConsoleDock.")
    }

    @objc private func startConsoleDock() {
        ConsoleDock.start(configuration: .sample)
        updateStatus("Started ConsoleDock.")
    }
}
